File: sdk/python/sdk.py
"""
Agent Observability SDK for Python agents.

Usage:
    from agent_observability import Client, LLMCall

    client = Client.from_env()

    with client.span("my-tool-call") as span:
        result = call_my_tool()
        span.set_attr("tool.name", "my-tool")

    with client.span("llm-call") as span:
        response = anthropic.messages.create(...)
        span.record_llm_call(LLMCall(
            model="claude-sonnet-4-20250514",
            prompt=prompt,
            completion=response.content[0].text,
            input_tokens=response.usage.input_tokens,
            output_tokens=response.usage.output_tokens,
        ))
"""

import ssl
import time
import uuid
import os
from contextlib import contextmanager
from dataclasses import dataclass, field
from typing import Optional, Iterator
import grpc


@dataclass
class LLMCall:
    model: str
    prompt: str
    completion: str
    input_tokens: int
    output_tokens: int
    cost_usd: float = 0.0


@dataclass
class Span:
    trace_id: str
    span_id: str
    parent_id: str
    name: str
    start_time: float
    attrs: dict = field(default_factory=dict)
    llm_call: Optional[LLMCall] = None
    _client: object = field(default=None, repr=False)

    def set_attr(self, key: str, value: str) -> "Span":
        self.attrs[key] = value
        return self

    def record_llm_call(self, call: LLMCall) -> "Span":
        self.llm_call = call
        return self

    def end(self) -> None:
        if self._client:
            self._client._flush(self, end_time=time.time())


class Client:
    """Thread-safe SDK client. Create once per process."""

    def __init__(self, gateway_addr: str, agent_name: str,
                 tls_cert: str, tls_key: str, tls_ca: str):
        self.agent_name = agent_name
        self._channel = self._build_channel(gateway_addr, tls_cert, tls_key, tls_ca)
        self._trace_id_var: Optional[str] = None
        self._span_id_var: Optional[str] = None

    @classmethod
    def from_env(cls) -> "Client":
        """Build a client from environment variables.

        Required env vars:
            OTEL_GATEWAY_ADDR  — e.g. gateway:4317
            OTEL_AGENT_NAME    — identifies this agent
            TLS_CERT           — path to client cert
            TLS_KEY            — path to client key
            TLS_CA             — path to CA cert
        """
        return cls(
            gateway_addr=os.environ["OTEL_GATEWAY_ADDR"],
            agent_name=os.environ["OTEL_AGENT_NAME"],
            tls_cert=os.environ["TLS_CERT"],
            tls_key=os.environ["TLS_KEY"],
            tls_ca=os.environ["TLS_CA"],
        )

    @contextmanager
    def span(self, name: str, trace_id: Optional[str] = None) -> Iterator[Span]:
        """Context manager that opens a span and ends it on exit.

        Example:
            with client.span("fetch-context") as s:
                s.set_attr("db.table", "documents")
                result = db.query(...)
        """
        tid = trace_id or self._trace_id_var or str(uuid.uuid4()).replace("-", "")
        parent = self._span_id_var
        sid = str(uuid.uuid4()).replace("-", "")

        s = Span(
            trace_id=tid,
            span_id=sid,
            parent_id=parent or "",
            name=name,
            start_time=time.time(),
            _client=self,
        )

        # Push context
        prev_trace, prev_span = self._trace_id_var, self._span_id_var
        self._trace_id_var = tid
        self._span_id_var = sid

        try:
            yield s
        finally:
            s.end()
            # Pop context
            self._trace_id_var = prev_trace
            self._span_id_var = prev_span

    def _flush(self, span: Span, end_time: float) -> None:
        """Serialize span to protobuf and send over gRPC. Stub."""
        pass

    def _build_channel(self, addr: str, cert: str, key: str, ca: str):
        with open(cert, "rb") as f:
            cert_data = f.read()
        with open(key, "rb") as f:
            key_data = f.read()
        with open(ca, "rb") as f:
            ca_data = f.read()

        creds = grpc.ssl_channel_credentials(
            root_certificates=ca_data,
            private_key=key_data,
            certificate_chain=cert_data,
        )
        return grpc.secure_channel(addr, creds)

    def close(self) -> None:
        self._channel.close()

    def __enter__(self):
        return self

    def __exit__(self, *_):
        self.close()

File: sdk/go/sdk.go
// Package sdk provides instrumentation for AI agents to emit telemetry
// to the observability gateway over mTLS/gRPC.
package sdk

import (
	"context"
	"crypto/tls"
	"time"

	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials"
)

// Client is the agent-side SDK. Create one per process.
type Client struct {
	conn      *grpc.ClientConn
	agentName string
}

// Config holds connection settings for the SDK.
// All fields map to environment variables for 12-factor compatibility.
type Config struct {
	GatewayAddr string // OTEL_GATEWAY_ADDR, e.g. "gateway:4317"
	AgentName   string // OTEL_AGENT_NAME — used as mTLS CommonName
	TLSCert     string // path to client cert
	TLSKey      string // path to client key
	TLSCACert   string // path to CA cert
}

// New creates an SDK client and establishes the mTLS connection to the gateway.
func New(cfg Config) (*Client, error) {
	cert, err := tls.LoadX509KeyPair(cfg.TLSCert, cfg.TLSKey)
	if err != nil {
		return nil, err
	}

	tlsCfg := &tls.Config{
		Certificates: []tls.Certificate{cert},
		MinVersion:   tls.VersionTLS13,
	}

	creds := credentials.NewTLS(tlsCfg)
	conn, err := grpc.NewClient(cfg.GatewayAddr, grpc.WithTransportCredentials(creds))
	if err != nil {
		return nil, err
	}

	return &Client{conn: conn, agentName: cfg.AgentName}, nil
}

// Span represents a unit of work in the agent execution
type Span struct {
	client    *Client
	traceID   string
	spanID    string
	parentID  string
	name      string
	startTime time.Time
	attrs     map[string]string
	llmCall   *LLMCall
}

// StartSpan begins a new span. Call span.End() when the work is done.
func (c *Client) StartSpan(ctx context.Context, name string) (*Span, context.Context) {
	s := &Span{
		client:    c,
		traceID:   traceIDFromContext(ctx),
		spanID:    newID(),
		parentID:  spanIDFromContext(ctx),
		name:      name,
		startTime: time.Now(),
		attrs:     make(map[string]string),
	}
	return s, contextWithSpan(ctx, s)
}

// SetAttr adds a key-value attribute to the span.
func (s *Span) SetAttr(key, value string) *Span {
	s.attrs[key] = value
	return s
}

// RecordLLMCall attaches LLM-specific metadata to this span.
// Call this after your LLM call completes.
func (s *Span) RecordLLMCall(call LLMCall) *Span {
	s.llmCall = &call
	return s
}

// End closes the span and flushes it to the gateway.
func (s *Span) End(ctx context.Context) {
	_ = s.client.flush(ctx, s, time.Now())
}

// LLMCall holds metadata about a single LLM inference call.
type LLMCall struct {
	Model        string
	Prompt       string
	Completion   string
	InputTokens  int32
	OutputTokens int32
	CostUSD      float64
}

func (c *Client) flush(ctx context.Context, s *Span, endTime time.Time) error {
	// Build and send OTLP span over gRPC — implementation calls generated proto client
	return nil // stub
}

// Close shuts down the gRPC connection gracefully.
func (c *Client) Close() error {
	return c.conn.Close()
}

func newID() string {
	// generate a random 16-byte hex ID
	return "" // stub — use crypto/rand in real impl
}

type contextKey int

const (
	traceIDKey contextKey = iota
	spanIDKey
	spanKey
)

func traceIDFromContext(ctx context.Context) string {
	if id, ok := ctx.Value(traceIDKey).(string); ok {
		return id
	}
	return newID()
}

func spanIDFromContext(ctx context.Context) string {
	if id, ok := ctx.Value(spanIDKey).(string); ok {
		return id
	}
	return ""
}

func contextWithSpan(ctx context.Context, s *Span) context.Context {
	ctx = context.WithValue(ctx, traceIDKey, s.traceID)
	ctx = context.WithValue(ctx, spanIDKey, s.spanID)
	return context.WithValue(ctx, spanKey, s)
}
